dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/managedobject/CDIManagedObjectService.java

```diff
@@ -36,76 +37,40 @@
     private CDIRuntime cdiRuntime;
 
-    public void activate(ComponentContext cc) {
-        cdiServiceRef.activate(cc);
-        defaultMOSRef.activate(cc);
-    }
-
-    public void deactivate(ComponentContext cc) {
-        cdiServiceRef.deactivate(cc);
-        defaultMOSRef.deactivate(cc);
-        cdiRuntime = null;
-    }
-
-    @Reference(name = "defaultManagedObjectService", service = DefaultManagedObjectService.class, policy = ReferencePolicy.DYNAMIC, cardinality = ReferenceCardinality.MANDATORY)
-    protected void setDefaultManagedObjectService(ServiceReference<DefaultManagedObjectService> ref) {
-        defaultMOSRef.setReference(ref);
-    }
-
-    protected void unsetDefaultManagedObjectService(ServiceReference<DefaultManagedObjectService> ref) {
-        defaultMOSRef.unsetReference(ref);
-    }
-
-    private DefaultManagedObjectService getDefaultManagedObjectService() {
-        return defaultMOSRef.getServiceWithException();
-    }
+    @Activate
+    public CDIManagedObjectService(@Reference DefaultManagedObjectService defaultMOS) {
 
-    @Reference(name = "cdiService", service = CDIService.class, policy = ReferencePolicy.DYNAMIC, cardinality = ReferenceCardinality.MANDATORY)
-    protected void setCDIService(ServiceReference<CDIService> ref) {
-        cdiServiceRef.setReference(ref);
+        this.defaultMOS = defaultMOS;
     }
 
-    protected void unsetCDIService(ServiceReference<CDIService> ref) {
-        cdiServiceRef.unsetReference(ref);
+    @Reference
+    protected void setCDIRuntime(CDIService cdiService) {
+        this.cdiRuntime = (CDIRuntime) cdiService;
     }
 
-    private CDIService getCDIService() {
-        return AccessController.doPrivileged(new PrivilegedAction<CDIService>() {
-            @Override
-            public CDIService run() {
-                return cdiServiceRef.getServiceWithException();
-            }
-        });
-
-    }
-
-    private CDIRuntime getCDIRuntime() {
-        if (this.cdiRuntime == null) {
-            this.cdiRuntime = (CDIRuntime) getCDIService();
-        }
-        return this.cdiRuntime;
+    //CDIRuntime can get deactivated after this object is created, so we need to track that and ensure we do not send calls towards it after it is deactivated.
+    protected void unsetCDIRuntime(CDIService cdiService) {
+        this.cdiRuntime = null;
     }
 
     @Override
     public <T> ManagedObjectFactory<T> createManagedObjectFactory(ModuleMetaData mmd, Class<T> klass,
                                                                   boolean requestManagingInjectionAndInterceptors) throws ManagedObjectException {
         if (!HttpJspPage.class.isAssignableFrom(klass) && isCDIEnabled(mmd)) {
-            return new CDIManagedObjectFactoryImpl<T>(klass, getCDIRuntime(), requestManagingInjectionAndInterceptors);
+            return new CDIManagedObjectFactoryImpl<T>(klass, cdiRuntime, requestManagingInjectionAndInterceptors);
```

Comment:
This pattern is in several places.  Now that cdiRunime is not final it is possible that it will get nulled out just after passing the `isCDIEnabled(mmd)` call above.  It is probably a really small window, but it is still a possible window where null would be passed to this constructor to `CDIManagedObjectFactoryImpl`